### DISCUSSION TITLE
RFC: Reduce macro footprint in sbt 2.0

### DISCUSSION THREAD

**Discussion by kubukoz** [sbt/sbt#7591] [Ideas]
Hi!

I've wanted to discuss this for a while but never found the focus to actually write it all down, so here it is :)

In typical sbt setups, users use a number of macros on the daily basis. These can have their own problems, among them are these:

- complicated error messages
- portability issues (a problem in the Scala 2 -> 3 migration only, AFAIK)
- difficulty in abstraction: building on top of macros requires making more macros

Some examples: the `project` macro, the `.value` macro on settings/tasks: these are the ones I'm familiar with the most, both as a user and a plugin maintainer.

I'm not going to argue against the usage of macros, because clearly there's been enough benefit in them to keep using them, but I think we can improve the status quo with a couple changes, while also not being too invasive on the users..

## smaller macros

This is mostly for things like `project` - instead of the whole method being a macro, we could use a model similar to [sourcecode](https://github.com/com-lihaoyi/sourcecode/):

```scala
def project(using name: sourcecode.Name) = Project(name.value, file(name.value))
```

`sourcecode.Name` could probably be replaced with an in-house implementation based on the existing macros / on Scala 3 sourcecode ([MIT licensed](https://github.com/com-lihaoyi/sourcecode/blob/main/LICENSE)).

I think this would also be usable in `KeyMacro` (`settingKey`, `taskKey`, `inputKey`) and possibly others. Is it a huge improvement? I don't know, but certainly it makes it easier to define custom utilities, such as a local `project`-like function that sets some default settings, validates naming conventions etc.

## alternative, well-documented DSLs

This has [been suggested before](https://github.com/sbt/sbt/discussions/7174#discussioncomment-5685981) (cc @julienrf): in addition to the `.value` / `.taskValue`-based DSL, we should have a visible and well-documented way to compose/derive settings/tasks with applicative/selective/monadic style, i.e. explicitly calling a `map` / `map2` / `mapN` (e.g. if someone were to implement Cats interop). I think these exist in some ways already, but in my past experience I've found these not-as-well-documented as the macros, and with some quirks (e.g. `map` changing the return type to something too wide).

---

Finally, I'll admit I'm not that familiar with the codebase, and these suggestions are coming mostly from a user's POV. I'm willing to invest time into implementing these suggestions, but will likely need some guidance along the way, and would like some feedback on the idea first, to avoid any wasted efforts.

Please let me know what you think!

**Comment by eed3si9n** [comment]
Hi. Thanks for taking the time to write this proposal.

Setting aside the fact that with sourcecode, the example is building on top of macros, I think this is a great suggestion, and I think it would improve the overall quality of the code base. I'm amazed at how elegant sourcecode-driven function has the equivalent power of macro-based implementation, but much simpler (My colleague @joan38 was showing some other sourcecode use today coincidentally). I think this would be a fun thing to explore.

>  we should have a visible and well-documented way to compose/derive settings/tasks with applicative/selective/monadic style, i.e. explicitly calling a `map` / `map2` / `mapN` (e.g. if someone were to implement Cats interop).

We did have that, and that's basically [sbt 0.12 documentation](https://www.scala-sbt.org/0.12.4/docs/Getting-Started/More-About-Settings.html). I think the challenge is that sbt's task graph is a heterogeneous list that contains a uniform `F1[_]` that forms the graph, whose parameter can either be `A` (pure value), or `F2[A]` (effect) that can either `mapN` or `flatMap`, so we have `F1[A1] :*: F1[F2[A2]] :*: F1[A3] :*: HNil` etc. Writing it out by hand is possible, but I don't think it would be fun or cognitively easy to get started. In fact we (sort of) know from history that people did not like that experience.

Maybe an interesting experiment might be to come up with a better abstraction over the above problem (heterogeneous `F1[_]`/`F1[F2[_]]`), like by using Eff or something. But given the resource constraint of sbt, I think we'd get better return-of-investment by focusing on things like remote cache, which are somewhat proven to work for other build tools.

**Reply by kubukoz** [reply]
> Setting aside the fact that with sourcecode, the example is building on top of macros

of course! My intention was just to have _one macro_ with more use-sites, rather than multiple macros that all essentially do one thing :)
